### ISSUE TITLE
The search algo needs improvement

### ISSUE THREAD

**Issue by esDotDev** [dart-lang/pub-dev#3994]
Currently the search as it is is not very helpful.

The main issue is that, there's no way to search by relevance, and also order by likes or popularity, but what is considered 'relevant' is far too broad.

Use case: I'm looking for plugins that help me format text. I want to see all the plugins that specifically mention text formatting, and order them by popularity.

So, I search for "Text Formatting"

[Results](https://pub.dev/packages?q=text+formatting):
- 1928 results off the top, that seems high for a fairly specific search, 
- Using the relevance view, I get what is seemingly a very random collection of results. 
  - Some of the top ones are highly voted, others have no votes, 
  - there seems to be very little rhyme or reason to the order here. 
  - The 5th overall result for "Text Formatting" is "Dolumns" with 0 likes, and 42% popularity. WAT?
- I know for a fact there are at least a dozen highly voted text formatting plugins out there, almost none of them are on the front page here.
- So, I switch to Overall Score, it doesn't seem to do anything
- So, I switch to order by "Most Likes" now there is maybe one actual Text Formatting lib in the first 50 pages, it's just filled with a collection of every top plugin in existence, provider, get, http etc... why is provider or http coming back at all in a search for "Text Formatting"? Neither of these libs have the word `formatting` anywhere in their title, desc or readme files.
* So I switch to order by Pub Score, but then I get 50 pages of arbitrarily ordered results, all with 110 pub score, so this is entirely useless as well

Some suggestions:
* Give us more control over what is searched. If we could specify title only, title + desc, etc, that would be very helpful.
* Additionally the simple abiity of and/or would go very far. Generally speaking, when I search for 2 words, I want results that contain both those words.
* Return less results so the various sorting options actually have value

**Comment by isoos** [comment]
Really great suggestions, thank you! I'll check if we have some low-hanging fruits somewhere...

**Comment by esDotDev** [comment]
Just to expand a bit, if I could for example search for "Text" in just the Title/Desc, and have only those 30 or 40 results returned, then all of the other sorting types would all have value and I could very quickly find all the top rated plugins that deal with text. 

One step further would be text || html || richtext, only return results that have one of those, and only check title/desc. Then I would most definitely get every single formatting lib out there in a really nice list and be able to choose ,y preferred one quickly and easily.

**Comment by isoos** [comment]
Closing this as we've deployed a new search on the site, with reduced fuzzy matching on the text contents, which addresses many of the points in this reported issue. Some reflection on the original points and parts that will probably remain the same:

> 1928 results off the top, that seems high for a fairly specific search

Without the fuzzy matching, it is reduced to 60.

> Using the relevance view, I get what is seemingly a very random collection of results.

It is a combination of pub points, text match relevance and popularity (downloads + likes). There is no good way to order or expose the weights in such case.

> The 5th overall result for "Text Formatting" is "Dolumns" with 0 likes, and 42% popularity. WAT?

In its description it says: `A Dart library for formatting text output into columns. [...]` I'd say it is a close and relevant text match. Some other results in the first page have the match only in their readme, which is ranked a bit lower than package name or description.

> One step further would be `text || html || formatting`

https://github.com/dart-lang/pub-dev/issues/1752
